### PR TITLE
PERF-#6753: Preserve dtypes cache on '.__setitem__()'

### LINKED ISSUE #6753: Preserve dtypes cache on `df[existing_col] = scalar`
It currently loses all dtypes on `.__setitem__()`
```python
import modin.pandas as pd

df = pd.DataFrame({"a": [1, 2, 3], "b": [3, 4, 5]})
df["b"] = 10

# known dtypes: {};
# cols with unknown dtypes: ['a', 'b'];
print(df._query_compiler._modin_frame._dtypes)
```

### PR DESCRIPTION
**Pull request by dchigarev**
## What do these changes do?

Detect the value's dtype and insert in into partial dtypes for the result of `.setitem()`

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [ ] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [ ] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [ ] signed commit with `git commit -s` 
- [ ] Resolves #6753 
- [ ] tests added and passing
- [ ] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/core/storage_formats/pandas/query_compiler.py [hunk 1]
```diff
@@ -2911,6 +2912,27 @@ def setitem_builder(df, internal_indices=[]):  # pragma: no cover
             idx = self.get_axis(axis ^ 1).get_indexer_for([key])[0]
             return self.insert_item(axis ^ 1, idx, value, how, replace=True)
 
+        if axis == 0:
+            if hasattr(value, "dtype"):
+                value_dtype = value.dtype
+            elif is_scalar(value):
+                value_dtype = np.dtype(type(value))
+            else:
+                value_dtype = np.array(value).dtype
```
Comment (anmyachev) [inline modin/core/storage_formats/pandas/query_compiler.py:2921]: It's probably better to move this into a separate function.

**Review by anmyachev — DISMISSED** [review]
LGTM!

File: modin/core/storage_formats/pandas/query_compiler.py (hunk 1 above)
Comment (dchigarev) [reply modin/core/storage_formats/pandas/query_compiler.py:2921]: will move for both `.insert()` and `__setitem()` after one of these PRs will be merged (#6757, #6758)

Comment (anmyachev) [reply modin/core/storage_formats/pandas/query_compiler.py:2929]: #6757 is merged

**Review by anmyachev — DISMISSED** [review]

**Comment by anmyachev** [comment]
@dchigarev CI is failed. Some problems with imports.

**Review by anmyachev — APPROVED** [review]

**Pull request merged**